src/allmydata/scripts/tahoe_daemonize.py

```diff
@@ -44,67 +63,70 @@ def getUsage(self, width=None):
         return t
 
-class StopOptions(BasedirOptions):
-    def parseArgs(self, basedir=None):
-        BasedirOptions.parseArgs(self, basedir)
-
-    def getSynopsis(self):
-        return ("Usage:  %s [global-options] stop [options] [NODEDIR]"
-                % (self.command_name,))
-
-class RestartOptions(StartOptions):
-    subcommand_name = "restart"
 
-class RunOptions(StartOptions):
-    subcommand_name = "run"
+class MyTwistdConfig(twistd.ServerOptions):
+    subCommands = [("DaemonizeTahoeNode", None, usage.Options, "node")]
 
 
-class MyTwistdConfig(twistd.ServerOptions):
-    subCommands = [("StartTahoeNode", None, usage.Options, "node")]
+class DaemonizeTheRealService(Service):
 
-class StartTahoeNodePlugin:
-    tapname = "tahoenode"
-    def __init__(self, nodetype, basedir):
+    def __init__(self, nodetype, basedir, options):
         self.nodetype = nodetype
         self.basedir = basedir
-    def makeService(self, so):
+
+    def startService(self):
         # delay this import as late as possible, to allow twistd's code to
         # accept --reactor= selection. N.B.: this can't actually work until
         # this file, and all the __init__.py files above it, also respect the
         # prohibition on importing anything that transitively imports
         # twisted.internet.reactor . That will take a lot of work.
-        if self.nodetype == "client":
-            from allmydata.client import Client
-            return Client(self.basedir)
-        if self.nodetype == "introducer":
-            from allmydata.introducer.server import IntroducerNode
-            return IntroducerNode(self.basedir)
-        if self.nodetype == "key-generator":
+
+        def key_generator_removed():
             raise ValueError("key-generator support removed, see #2783")
-        if self.nodetype == "stats-gatherer":
-            from allmydata.stats import StatsGathererService
-            return StatsGathererService(verbose=True)
-        raise ValueError("unknown nodetype %s" % self.nodetype)
 
-def identify_node_type(basedir):
-    for fn in listdir_unicode(basedir):
-        if fn.endswith(u".tac"):
-            tac = str(fn)
-            break
-    else:
-        return None
+        def start():
+            node_to_instance = {
+                u"client": lambda: namedAny("allmydata.client.Client")(self.basedir),
+                u"introducer": lambda: namedAny("allmydata.introducer.server.IntroducerNode")(self.basedir),
+                u"stats-gatherer": lambda: namedAny("allmydata.stats.StatsGathererService")(self.basedir, verbose=True),
+                u"key-generator": key_generator_removed,
+            }
 
-    for t in ("client", "introducer", "key-generator", "stats-gatherer"):
-        if t in tac:
-            return t
-    return None
+            try:
+                service_factory = node_to_instance[self.nodetype]
+            except KeyError:
+                raise ValueError("unknown nodetype %s" % self.nodetype)
```

Comment:
Review: mmm, probably annoying.

But I like this kind of PR when there are code-base-wide problems, because when you're rebasing/merging any older branches it's way easier to know what the right way to resolve the conflicts is ("change everything to .format(), okay").